#!/usr/bin/env python
import os
import sys

import urllib3
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

import logging
log = logging.getLogger("redmine")
log.setLevel(logging.DEBUG)
log.addHandler(logging.StreamHandler())

from redminelib import Redmine
from settings import *

def connect():
    log.debug(f"Connecting to: {REDMINE_URL}")
    redmine = Redmine(REDMINE_URL, key=REDMINE_KEY, requests={'verify': False})

    user = redmine.user.get("current")
    log.info(f"Hi, {user.firstname} {user.lastname}")
    return redmine

def download_attached(redmine, issue_id):
    issue = redmine.issue.get(issue_id)
    log.debug(issue)
    log.debug(f"Attachments: {[a.filename for a in issue.attachments]}")
    download_dir = f"{DOWNLOAD_ROOT}/{issue_id}/"
    os.makedirs(download_dir,  exist_ok=True)

    log.info("Start downloading...")
    for a in issue.attachments:
        log.info(f"\t* {a.filename}")
        a.download(savepath=download_dir)

    log.info(f"Successfully downloaded {len(issue.attachments)} attachments to {download_dir}")

if __name__ == "__main__":
    if len(sys.argv) != 2:
        log.error("Wrong number of arguments!")
        log.info(f"Usage {__file__} issue_id")
        sys.exit(1)
    if not REDMINE_KEY:
        log.error("REDMINE_KEY Not defined! Set it in settings.py")
        sys.exit(1)

    download_attached(connect(), sys.argv[1])
